### PR TITLE
The real slips rewrite

### PR DESCRIPTION
**Pull request by Cyberboss**
Added a datum for trying to force a mob to move to a point over some period of time

Slipping and meathooking now use /datum/forced_movement which is more refined than the shitty for-spawn that slips used to use and the instant teleportation of the meat hook

Also fixed a bug in forceMove that caused Entered and Exited to be called even if the atom was moved to the same loc

Also fixes an unreported bug I found with freon
🆑 Cyberboss
fix: Freon slips no longer trap you in an endlessly reversing cycle
/🆑

- [x] Add callbacks once #22476 is merged

### CODE REVIEW

File: code/datums/forced_movement.dm [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	if(steps_to_take)
+		for(var/i = 0, i < steps_to_take, ++i)
+			if(!TryMove())
+				victim.force_moving = null
+				victim.forceMove(victim.loc)	//get the side effects of moving here that require us to currently not be force_moving aka reslipping on ice
+				qdel(src)
+				return
+		last_processed = world.time
+
+/datum/forced_movement/proc/TryMove(recursive = FALSE)
```
Comment (Cyberboss) [inline code/datums/forced_movement.dm]: Any better way to do diagonal movement?

File: code/datums/forced_movement.dm [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	var/atom/target
+	var/stunned
+	var/last_processed
+	var/steps_per_tick
+	var/moved_y_last
+	var/allow_tabling
+	var/spin
+															//as fast as ssfastprocess
+/datum/forced_movement/New(atom/movable/victim, atom/target, steps_per_tick = 0.5, allow_tabling = FALSE, spin = FALSE)
+	src.victim = victim
```
Comment (ChangelingRain) [inline code/datums/forced_movement.dm]: avoid having to do src.var by renaming the arguments for New()

File: code/datums/forced_movement.dm (hunk 1 above)
Comment (ChangelingRain) [inline code/datums/forced_movement.dm]: Couldn't you just step_towards(vic, tar)?

File: code/datums/forced_movement.dm [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	STOP_PROCESSING(SSfastprocess, src)
+	return ..()
+
+/datum/forced_movement/process()
+	if(qdeleted(victim) || !victim.loc || qdeleted(target) || !target.loc)
+		qdel(src)
+		return
+	var/steps_to_take = round(steps_per_tick * (world.time - last_processed))
+	if(steps_to_take)
+		for(var/i = 0, i < steps_to_take, ++i)
```
Comment (Iamgoofball) [inline code/datums/forced_movement.dm]: Nope, wrong type of loop

**Review by Iamgoofball — CHANGES_REQUESTED** [review]

File: code/game/objects/structures.dm [hunk 4]
```diff
@@ -64,6 +64,11 @@
 		step(O, get_dir(O, src))
 	return
 
+/obj/structure/proc/do_climb(atom/movable/A)
+	density = 0
```
Comment (ChangelingRain) [inline code/game/objects/structures.dm]: Are there any climbable objects with a density of 0? You should at least account for them

Comment (Cyberboss) [inline code/game/objects/structures.dm]: This is just a refactor of /obj/structures climb code so forced_movement can use it. Should be fine if it was before

**Comment by optimumtact** [comment]
Why not use the same subsystem as spacemovement, or roll them together?

**Comment by Cyberboss** [comment]
This does some very basic obstacle avoidance. Also it's a lot faster than
space drift.

**Comment by Armhulen** [comment]
wher the fuck id it?

**Comment by Iamgoofball** [comment]
IT'S HERE

File: code/datums/forced_movement.dm (hunk 1 above)
Comment (MrPerson) [inline code/datums/forced_movement.dm]: Lots of ways to do this: See shooting for an example that's actually straight

**Comment by MrStonedOne** [comment]
You know, you might be able to merge throwing into this... meaning you could absorb my pr.

It seems like it could be expanded enough to also do space drift, using a subtype of forced move.

So could throwing.

I like this system much better than my low effort merge of throw_at into a subsystem.

File: code/datums/forced_movement.dm [hunk 5]
```diff
@@ -0,0 +1,98 @@
+//Just new and forget
+/datum/forced_movement
+	var/atom/movable/victim
+	var/atom/target
+	var/last_processed
+	var/steps_per_tick
+	var/allow_tabling
+	var/spin
+															//as fast as ssfastprocess
+/datum/forced_movement/New(atom/movable/avictim, atom/atarget, asteps_per_tick = 0.5, aallow_tabling = FALSE, aspin = FALSE)
```
Comment (MrStonedOne) [inline code/datums/forced_movement.dm]: I must say, i'd really rather prefer something like src.victim = avictim. adding a to the front just makes it harder to read at a glance.

up to you.

Comment (optimumtact) [inline code/datums/forced_movement.dm]: use a _

Comment (optimumtact) [inline code/datums/forced_movement.dm]: >aallow_tabling
aaaaaaaa

Comment (RemieRichards) [inline code/datums/forced_movement.dm]: yeah :+1: to what mso said.

File: code/datums/forced_movement.dm [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+				if(recursive)
+					return FALSE
+				else
+					. = TryMove(TRUE)
+
+	. = . && vic.loc != tar.loc
+
+/mob/Bump(atom/A)
+	. = ..()
+	if(force_moving && force_moving.allow_tabling && istype(A,/obj/structure))
```
Comment (RemieRichards) [inline code/datums/forced_movement.dm]: should rename ```allow_tabling``` to ```allow_climbing``` or ```do_climbing``` or something.

File: code/datums/forced_movement.dm [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+/datum/forced_movement/proc/TryMove(recursive = FALSE)
+	var/atom/movable/vic = victim	//sanic
+	var/atom/tar = target
+
+	if(!recursive)
+		if(iscarbon(vic))
+			var/mob/living/carbon/C = vic
+			if(spin)
+				C.spin(1,1)
```
Comment (RemieRichards) [inline code/datums/forced_movement.dm]: make this a movable proc, ```dirSpin()```, and then generalise the use here to movable.

File: code/datums/forced_movement.dm [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+				if(step(vic, SOUTH))
+					. = TRUE
+
+			if(!.)
+				if(recursive)
+					return FALSE
+				else
+					. = TryMove(TRUE)
+
+	. = . && vic.loc != tar.loc
```
Comment (RemieRichards) [inline code/datums/forced_movement.dm]: paranthesis, for readability.

File: code/datums/forced_movement.dm [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+	target = atarget
+	steps_per_tick = asteps_per_tick
+	allow_tabling = aallow_tabling
+	spin = aspin
+
+	last_processed = world.time
+
+	. = ..()
+
+	if(!avictim.force_moving)
```
Comment (RemieRichards) [inline code/datums/forced_movement.dm]: you can't replace an existing ```force_movement```?

**Comment by RemieRichards** [comment]
I don't see these getting cleaned up anywhere, are you supposed to hold a lingering ref to a finished forcemove?

File: code/datums/forced_movement.dm (hunk 5 above)
Comment (ChangelingRain) [inline code/datums/forced_movement.dm]: or use new_ as a prefix

Comment (Cyberboss) [inline code/datums/forced_movement.dm]: >opinions

File: code/datums/forced_movement.dm (hunk 7 above)
Comment (Cyberboss) [inline code/datums/forced_movement.dm]: I'll just replace this with a per-step callback once those are in

**Comment by Cyberboss** [comment]
You can qdel it to stop it early or it will qdel itself when it finishes/can't go further. It handles its own references

**Comment by Cyberboss** [comment]
Is there a good reason SSfastprocess isn't 1 tick?

**Comment by optimumtact** [comment]
Well it's not supposed to be that fast, just faster than the normal obj subsystem.

But the more things you put on it the slower it will get.

**Comment by Iamgoofball** [comment]
Fast process is like fastmos. It runs as fast as possible.

**Comment by optimumtact** [comment]
>wait = 2

**Comment by Iamgoofball** [comment]
It doesn't use that wait.

**Comment by MrStonedOne** [comment]
Is there a good reason SSfastprocess isn't 1 tick?

Ticker subsystems run first every tick. If it ran every tick and called something that didn't tick check then the mc wouldn't be able to do anything else

**Comment by optimumtact** [comment]
Callbacks?

**Comment by Cyberboss** [comment]
Yeah I'll take care of it when I get home

**Comment by Cyberboss** [comment]
Testmerbeg

**Comment by optimumtact** [comment]
Will chase for test merges

**Comment by PKPenguin321** [comment]
hahahaha

**Comment by optimumtact** [comment]
Did this get test merged?

**Comment by Cyberboss** [comment]
It has been for a while now

**Pull request merged**